Added driver support for m24256 eeprom

Arunmanialagarsamy opened the pull request:
Added the driver support for m24256 eeprom. It allows to read and write the data

ArunmaniAlagarsamy2710 committed:
boards: st: b_u585i_iot02a: Add eeprom child node in i2c2

Added eeprom devicetree support for b_u585i_iot02a board.

Signed-off-by: Arunmani Alagarsamy <arunmani.a@capgemini.com>

Arunmanialagarsamy commented:
@henrikbrixandersen Please take a review

henrikbrixandersen commented:
> @henrikbrixandersen Please take a review

I will get to it, you only submitted this 4 days ago. While you wait, I suggest you read up on our documented Contributor Expectations, especially the section on review escalation: https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#pr-review-escalation

Arunmanialagarsamy commented:
@henrikbrixandersen Waiting for your approval, please take a look.

kartben commented on samples/drivers/eeprom/boards/b_u585i_iot02a.overlay:
```diff
@@ -0,0 +1,14 @@
+		 compatible = "st,m24256";
+		 reg = <0x56>;
+		 status = "okay";
+		 size = <3600>;
```
The datasheet for this board suggests it's 256Kbits?

kartben commented on samples/drivers/eeprom/boards/b_u585i_iot02a.overlay:
Don't make this an overlay and directly add to the board's DTS instead. You may also update the. Boards readme accordingly.

kartben commented on dts/bindings/mtd/st,m24256.yaml:
```diff
@@ -0,0 +1,17 @@
+
+description: st m24256 series I2C EEPROM
+
+compatible: "st,m24256"
```
I am guessing this could be a generic st,m24xxx driver, no? i.e. would a m24512 driver really be different?
Interestingly, `boards/shields/x_nucleo_eeprma2` already uses an `st,m24xxx` compatible :)

henrikbrixandersen replied:
Perhaps this can be made even more generic. From a quick glance, it looks very similar to https://github.com/zephyrproject-rtos/zephyr/pull/69327

benediktibk replied:
Just as a hint, M24M02 seems to be different though:
![image](https://github.com/zephyrproject-rtos/zephyr/assets/9367089/10895cb8-fd60-45b0-bddc-df19be744406)

It uses part of the I2C address for the memory address and therefore one I2C slave uses up multiple I2C addresses on the bus.

So probably we shouldn't be too catch-all with the compatible string.

henrikbrixandersen reviewed (changes requested):
There are so many similarities with the driver submitted in #69327 that I think these can be combined into one driver, supporting this "class" of I2C EEPROMs. See example in `drivers/eeprom/eeprom_at2x.c` which supports both Atmel AT24 and AT25 series (and compatibles) with one driver.

Please coordinate internally between the two submitters on how to proceed with one, common driver. I'll gladly participate in a discussion about this, either on the Zephyr Discord server, in an Architecture WG meeting or elsewhere.

str4t0m reviewed (changes requested):
This eeprom should already work with the `atmel,at24` driver as is.
Could you please give it a try on your b_u585i_iot02a board.

ArunmaniAlagarsamy2710 committed:
drivers: eeprom: Add driver support for m24xxx

Added the driver support for m24xxx eeprom. It allows to read and write the data

Signed-off-by: Arunmani Alagarsamy <arunmani.a@capgemini.com>

henrikbrixandersen reviewed (changes requested):
The st,m24xxx EEPROM family is already supported by the existing `eeprom_at2x.c` driver. Please see the following shield for how to describe these via devicetree:

https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/shields/x_nucleo_eeprma2/x_nucleo_eeprma2.overlay

Arunmanialagarsamy commented:
Lets close this PR as per our discussion
https://discordapp.com/channels/720317445772017664/1218200905044721765/1219957042874155008
